internal/resolver/delegatingresolver/delegatingresolver.go

```diff
@@ -0,0 +1,236 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+// Package delegatingresolver defines a resolver that can handle both target URI
+// and proxy address resolution, unless:
+//   - A custom dialer is set using WithContextDialer dialoption.
+//   - Proxy usage is explicitly disabled using WithNoProxy dialoption.
+//   - Client-side resolution is explicitly enforced using WithTargetResolutionEnabled.
+package delegatingresolver
+
+import (
+	"fmt"
+	"net/http"
+	"net/url"
+	"sync"
+
+	"google.golang.org/grpc/grpclog"
+	"google.golang.org/grpc/resolver"
+	"google.golang.org/grpc/serviceconfig"
+)
+
+var (
+	// HTTPSProxyFromEnvironment will be used and overwritten in the tests.
+	HTTPSProxyFromEnvironment = http.ProxyFromEnvironment
+
+	logger = grpclog.Component("delegating-resolver")
+)
+
+// delegatingResolver implements the `resolver.Resolver` interface. It uses child
+// resolvers for the target and proxy resolution. It acts as an intermediatery
+// between the child resolvers and the gRPC ClientConn.
+type delegatingResolver struct {
+	target         resolver.Target     // parsed target URI to be resolved
+	cc             resolver.ClientConn // gRPC ClientConn
+	targetResolver resolver.Resolver   // resolver for the target URI, based on its scheme
+	proxyResolver  resolver.Resolver   // resolver for the proxy URI; nil if no proxy is configured
+	targetAddrs    []resolver.Address  // resolved or unresolved target addresses, depending on proxy configuration
+	proxyAddrs     []resolver.Address  // resolved proxy addresses; empty if no proxy is configured
+	proxyURL       *url.URL            // proxy URL, derived from proxy environment and target
+	mu             sync.Mutex          // protects access to the resolver state during updates
+}
+
+func parsedURLForProxy(address string) (*url.URL, error) {
+	req := &http.Request{URL: &url.URL{Scheme: "https", Host: address}}
+	url, err := HTTPSProxyFromEnvironment(req)
+	if err != nil {
+		return nil, err
+	}
+	return url, nil
+}
+
+// New creates a new delegating resolver that is used to call the target and
+// proxy child resolver. If proxy is configured and target endpoint points
+// correctly points to proxy, both proxy and target resolvers are used else
+// only target resolver is used.
+//
+// For target resolver, if scheme is dns and target resolution is not enabled,
+// it stores unresolved target address, bypassing target resolution at the
+// client and resolution happens at the proxy server otherwise it resolves
+// and store the resolved address.
+//
+// It returns error if proxy is configured but proxy target doesn't parse to
+// correct url or if target resolution at client fails.
+func New(target resolver.Target, cc resolver.ClientConn, opts resolver.BuildOptions, targetResolverBuilder resolver.Builder, targetResolutionEnabled bool) (resolver.Resolver, error) {
```

Comment:
Can this be a `resolver.Builder.Build` method instead?  That would theoretically allow us to just substitute it into the channel instead of the `Builder` we get out of the resolver registry.

I think only `targetResolutionEnabled` is needed beyond that interface, but we would just not use this if `targetResolutionEnabled == false`, and use the registry's entry directly instead.

To find `targetResolverBuilder`, we'd do `resolver.Get` just like the channel does today.

Reply:
I dont think that is possible becuase:
- We cannot use registry's entry for `targetResolutionEnabled` because it only makes sense when used with proxy , and the proxy address is only introduced in `delegatingResolver` . Using just the target resolver will never be correct to connect to proxy.
- In addition to `resolver.Get` we also do a search for the target resolver in the dialOptions, for a custom resolver. If we want to use a custom resolver with proxy (like we want to do for the case 2) , we will need to pass the dialOptions if not the `targetResolverBuilder`.

Also, `gracefulSwitch` does something similar and does not have a `Build` method, but a `NewBalancer` method instead.

Reply:
> We cannot use registry's entry for targetResolutionEnabled because it only makes sense when used with proxy

Sorry, I'm not following this.

> In addition to resolver.Get we also do a search for the target resolver in the dialOptions

So we could do this if we passed in the resolvers from the dial options, or provided a way to ask for the resolvers from the channel (via ClientConn)?  This might be worth doing to keep the channel code simpler when integrating this delegating resolver into it.  We should discuss options.

> Also, gracefulSwitch does something similar and does not have a Build method, but a NewBalancer method instead.

This is a bad example.  We only recently decided to make the gsb a real balancer -- before that it required special operations.  Probably a good next step would be to make it implement `Build` with the same API.

Generally, we should look at other examples inside our code base for consistency where consistency is important and relevant.  Here we can be more consistent by matching the resolver API, vs. following something unrelated with a different background.

Reply:
> Sorry, I'm not following this.

What I meant to say was that we need the value of `TargetResolutionEnabled` because its value only matters when used along with proxy configurations, otherwise we just do the target resolution using the normal way (just the target resolver, no delegating resolver)

> So we could do this if we passed in the resolvers from the dial options, or provided a way to ask for the resolvers from the channel (via ClientConn)? This might be worth doing to keep the channel code simpler when integrating this delegating resolver into it. We should discuss options.

Yes, so passing the dialOptions or resolvers from dialOptions to the delegating resolver will also violate the parameters needed for build.